r/R/dplyr.R

```diff
@@ -539,6 +537,13 @@ mutate.arrow_dplyr_query <- function(.data,
     if (inherits(results[[new_var]], "try-error")) {
       msg <- paste('Expression', as_label(exprs[[i]]), 'not supported in Arrow')
       return(abandon_ship(call, .data, msg))
+    } else if (is_dataset && !inherits(results[[new_var]], "Expression")) {
+      # We need some wrapping to handle literal values
+      if (length(results[[new_var]]) != 1) {
+        msg <- paste0('In ', new_var, " = ", as_label(exprs[[i]]), ", only values of size one are recycled")
```

Comment:
If you `mutate(NULL)` or `transmute(NULL)` it gets caught by this and gives the message:
```
Error: In NULL = NULL, only values of size one are recycled
Call collect() first to pull data into R. 
```
We should handle that case separately and make it behave the same as a no-argument `mutate()` or `transmute()` (that is, return all the columns or no columns respectively)

Reply:
Why would one ever call `transmute()` or `mutate()` like this (on purpose)?

Reply:
P.S. named `NULL` should behave the same, e.g. `mutate(x = NULL, y = NULL, z = NULL)` should do the same as `mutate()`

Reply:
If there is an issue with NULL handling (or anything) that applies both to Table/RecordBatch and Dataset, can we make a followup JIRA for it?

Reply:
> Why would one ever call `transmute()` or `mutate()` like this (on purpose)?

An increasing amount of dplyr code is programmatically generated, not hand-written. It's quite common to have empty lists and `NULL` values getting passed to verbs in such cases.

Also, this could happen by accident if a variable has a `NULL` value and then it's used in the `mutate()` expression.

Reply:
Oh, and also you can use `mutate(existing_column = NULL)` to remove a column in dplyr

Reply:
I agree that we should error gracefully, and `mutate(existing_column = NULL)` seems like a valid thing to support. I expect this would just work with a Table/RecordBatch now, just not for a dataset with the extra scalar handling I added.

Reply:
The follow-up JIRA for empty/null in `transmute()` is ARROW-11880. I believe that's the only issue here that's not specific to Datasets.

Reply:
Yes, `mutate(existing_column = NULL)` does work for Table/RecordBatch now.